tools/autograd/gen_variable_type.py

```diff
@@ -67,6 +67,18 @@
 auto${ref} ${arg_name}_ = unpack${suffix}(${arg_name}, "${arg_name}", ${arg_pos});""")
 
 FUNCTION_DECLARATION = CodeTemplate("""\
+struct ${op} : public Function {
+  using Function::Function;
```

Comment:
There's a lot of repetition between these two snippets. Maybe jut make the superclass a template parameter and generate it somewhere in the code?